support for multiple datasets

jo-pol opened the pull request:
Changed command line options:
- multiple ways to specify multiple datasets: [-q <fedora-query>... | -i <dataset-id>... | --file <text-file>]
- without a system variable for the home folder, the working directory is used to locate the cfg/application.properties file, hard coded defaults provide a final fall back
- explicit shorthand characters avoid conflicts
- see README for more

also:
- a unit test (ConfSpec/printHelp) prints the synopsys/arguments info for the README
- refactoring settings and making it implicit makes method signatures cleaner.
- less verbose not found errors
- close input streams when reading datastreams. This happens in a loop, half a dozen of not closed streams per process does not matter much, but now it does matter.
- case classes for ProviderImpl and overriding toString for FedorClient simplifies logging while making it more informative

jo-pol committed:
Support for multiple datasets

Changed command line options:
- multiple ways to specify multiple datasets: [-q <fedora-query>... | -i <dataset-id>... | --file <text-file>]
- former --debug option turned into --apply-updates to prevent accidental destructive tests
- without a system variable for the home folder, the working directory is used to locate the cfg/application.properties file, hard coded defaults provide a final fall back
- explicit shorthand characters avoid conflicts
- see also README.md

also:
- a unit test prints the synopsys/arguments info for the README (ConfSpec/printHelp)
- refactoring settings and making it implicit makes method signatures cleaner.
- less verbose not found errors
- close input streams when reading datastrems. This happens in a loop, half a dozen of not closed streams per process does not matter much, but now it does matter.
- case classes for ProviderImpl and overriding toString for FedorClient simplifies logging while making it more informative

janvanmansum commented on README.md:
```diff
@@ -18,15 +19,34 @@ in the Fedora Commons Repository at the dataset level. This command extracts the
-* ``-s``, ``--solr-server-request-handler`` -- URL of the EASY SOLR Search Index server request handler
-* ``-x``, ``--output-xml-dir`` -- directory to write the documents to be sent to SOLR to
-* ``<easy-datataset-pid>...`` -- one or more dataset PIDs 
+      -a, --apply-updates               If omitted: only generate document(s), do not
```
I disagree with this change. Please revers it. For tasks that run the risk to mess up the actual data or metadata that may be a good thing. This command however just updates the search index. Would be annoying to have to specify everytime that you really mean to do it.

janvanmansum commented on README.md:
```diff
@@ -18,15 +19,34 @@ in the Fedora Commons Repository at the dataset level. This command extracts the
-* ``<easy-datataset-pid>...`` -- one or more dataset PIDs 
+      -a, --apply-updates               If omitted: only generate document(s), do not
+                                        send anything to SOLR
+      -b, --dataset-batch-size  <arg>   Number of datastes to read at once from the
```
typo in "datasets"

jo-pol committed:
Support for multiple datasets (review comments)

- typo in datastes
- reverted a command line option:
  --apply-updates back to --debug
  accidental index updates don't hurt

janvanmansum commented on src/main/scala/nl/knaw/dans/easy/solr/FedoraProviderImpl.scala:
```diff
@@ -18,27 +18,54 @@ package nl.knaw.dans.easy.solr
+    }
   }
 
+  /** Reads the Dublin Core meta data */
```
The comments are unnecessary. The method names use the familiar abbreviations, use elsewhere in EASY.

janvanmansum commented on src/test/scala/nl/knaw/dans/easy/solr/ConfSpec.scala:
```diff
@@ -0,0 +1,57 @@
+    result should include("${project.version}") // FIXME ?
+  }
+
+  it should "be pasted into README.md" in {
```
While creative, I see little value in providing this "unit test". Copy-pasting the output of the `--help` option is a neat trick but can just as well be done by running the application and copy-pasting from the terminal screen. Let's use unit test for what they intended for.

jo-pol replied:
I am too lazy for the build-unzip-run cycle, or create run configurations in the IDE which I loose frequently. This way I don't need to remember anything. Now I can just run the build or the tests to get it. During development I frequently had second thoughts on the options and your review comment made me change it again. So I copy-pasted it also frequently.

Seeing it between the test results also reminds me I should copy-paste it!

janvanmansum commented on src/test/scala/nl/knaw/dans/easy/solr/ConfSpec.scala:
```diff
@@ -0,0 +1,57 @@
+
+class ConfSpec extends FlatSpec with Matchers {
+
+  "printHelp" should "produce help info with values from custom properties file" in {
```
Is it useful to test this? It is really Scallop-functionality and not our own logic that you are testing here.

jo-pol replied:
This test does not go far enough: it should read the properties file and assert all the values appear as defaults. The keys of the test version should also match the keys in the assembly version of the properties files. Thus you assert everything gets properly maintained in the future.

janvanmansum commented on src/main/assembly/dist/bin/help.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/sh
```
I'm really in favour of "mean and lean" whenever possible. This script is obviously only there to provide some alternative way of starting the application in the development environment. However, there is a way to start it that does not include "for development only" customizations of the project.

That said, the trick with `dirname $0` is rather neat. I've seen it in your code before and wondered if we could make it the standard way, as it would obviate the use of the environment variable.

jo-pol replied:
> it would obviate the use of the environment variable.

I would really appreciate that

janvanmansum replied:
Continuing my previous remark (I was called away :), earlier I though there was a conflict between the following three requirements:
- The working directory should stay the same. This would _not_ be the case when using a strategy where you just `pushd` to the home-directory and `popd` afterwards. Not having this, however, is confusing when the command line application may take a file name (and therefore relative path) as an argument. Not all applications do, but it is sufficiently common to make this a general principle.
- The main jar can find its library dependencies. Currently this is achieved through relative paths in the main jar's `MANIFEST.MF` file.
- The application should have some reference to its "home", in order to be able to load configuration settings, etc.

It seems I was wrong about them being conflicting. The `MANIFEST.MF` file seems to resolve the paths to the libraries relative to the jar it is contained in, _not_ relative to the working directory (which I previously thought). The home-directory can be passed to the application as a Java system property.
